### PR TITLE
fix: new exceptions and tests

### PR DESCRIPTION
**Pull request by sattvikc**
## Summary of change

(A few sentences about this PR)

## Related issues

- Link to issue1 here
- Link to issue1 here

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your
changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here
highlighting the necessary changes)

## Checklist for important updates

- [ ] Changelog has been updated
    - [ ] If there are any db schema changes, mention those changes clearly
- [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
- [ ] `pluginInterfaceSupported.json` file has been updated (if needed)
- [ ] Changes to the version if needed
    - In `build.gradle`
- [ ] If added a new paid feature, edit the `getPaidFeatureStats` function in FeatureFlag.java file
- [ ] Had installed and ran the pre-commit hook
- [ ] If there are new dependencies that have been added in `build.gradle`, please make sure to add them
  in `implementationDependencies.json`.
- [ ] Issue this PR against the latest non released version branch.
    - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the
      latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    - If no such branch exists, then create one from the latest released branch.

## Remaining TODOs for this PR

- [ ] Item1
- [ ] Item2

### CODE REVIEW

File: src/test/java/io/supertokens/test/multitenant/StorageLayerTest.java [hunk 1]
```diff
@@ -62,4 +75,1075 @@ public void normalConfigContinuesToWork() throws InterruptedException, IOExcepti
...
+                                                    new ThirdPartyConfig.UserInfoMapKeyValue("id1", "email1", "email_verified1"),
+                                                    new ThirdPartyConfig.UserInfoMapKeyValue("id2", "email2", "email_verified2")
+                                            )
+                                    )
+                            }),
+                            new PasswordlessConfig(true),
+                            new JsonObject()
+                    ));
+                } catch (Exception e) {
+                    if (e.getMessage().toLowerCase().contains("the transaction might succeed if retried")) {
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/multitenant/StorageLayerTest.java:1101]:
```suggestion
```

File: src/test/java/io/supertokens/test/multitenant/StorageLayerTest.java [hunk 2]
```diff
@@ -62,4 +75,1075 @@ public void normalConfigContinuesToWork() throws InterruptedException, IOExcepti
...
+                        pass.set(false);
+                    }
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/multitenant/StorageLayerTest.java:1103]:
```suggestion
```

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/test/java/io/supertokens/test/multitenant/StorageLayerTest.java (hunk 1 above)
Comment (sattvikc) [reply src/test/java/io/supertokens/test/multitenant/StorageLayerTest.java]: Have moved this test to postgres plugin.

File: src/test/java/io/supertokens/test/multitenant/StorageLayerTest.java (hunk 2 above)
Comment (sattvikc) [reply src/test/java/io/supertokens/test/multitenant/StorageLayerTest.java]: Have moved this test to postgres plugin.

File: src/test/java/io/supertokens/test/multitenant/StorageLayerTest.java (hunk 1 above)
Comment (rishabhpoddar) [reply src/test/java/io/supertokens/test/multitenant/StorageLayerTest.java]: This test needs to be in the core only.

File: src/test/java/io/supertokens/test/multitenant/ConfigTest.java [hunk 3]
```diff
@@ -358,6 +366,538 @@ public void testDifferentWaysToGetConfigBasedOnConnectionURIAndTenantId()
...
 
+        process.kill();
+        assertNotNull(process.checkOrWaitForEvent(ProcessState.PROCESS_STATE.STOPPED));
+    }
+
+    @Test
+    public void testMappingSameUserPoolToDifferentConnectionURIThrowsError()
+            throws InterruptedException, IOException, TenantOrAppNotFoundException {
+        String[] args = {"../"};
+
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/multitenant/ConfigTest.java:377]: add a new test that will randomly generate a tenant:
- for each tenant config's value, you can specify a generateValue function, and expected outcome -> 
   - thirdpartId 
      - generate function will return a random string (UUID) -> expected outcome is success (UUID)
      -  generate function will return a null  -> expected outcome is InvalidConfigException
   - infomap
      - generate function will return null -> outcome is empty map

- You can use java reflection -> enforce that each config that we have is tested.

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/thirdparty/ThirdParty.java [hunk 4]
```diff
@@ -228,7 +229,33 @@ public static UserInfo[] getUsersByEmail(TenantIdentifier tenantIdentifier, Main
 
     public static void verifyThirdPartyProvidersArray(ThirdPartyConfig.Provider[] providers)
             throws InvalidProviderConfigException {
-        // TODO:
+        for (ThirdPartyConfig.Provider provider: providers) {
+            verifyThirdPartyProvider(provider);
+        }
+    }
+
+    public static void verifyThirdPartyProvider(ThirdPartyConfig.Provider provider)
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/thirdparty/ThirdParty.java:237]:
```suggestion
    private static void verifyThirdPartyProvider(ThirdPartyConfig.Provider provider)
```

File: src/main/java/io/supertokens/thirdparty/ThirdParty.java [hunk 5]
```diff
@@ -228,7 +229,33 @@ public static UserInfo[] getUsersByEmail(TenantIdentifier tenantIdentifier, Main
...
+        if (provider.thirdPartyId == null) {
+            throw new InvalidProviderConfigException("thirdPartyId cannot be null");
+        }
+
+        for (ThirdPartyConfig.ProviderClient client : provider.clients) {
+            verifyThirdPartyProviderClient(client, provider.thirdPartyId);
+        }
+    }
+
+    public static void verifyThirdPartyProviderClient(ThirdPartyConfig.ProviderClient client, String thirdPartyId)
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/thirdparty/ThirdParty.java:249]:
```suggestion
    private static void verifyThirdPartyProviderClient(ThirdPartyConfig.ProviderClient client, String thirdPartyId)
```

File: src/test/java/io/supertokens/test/multitenant/RandomConfigTest.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        int notOkCount = 0;
+
+        for (int i = 0; i < 10000; i++) {
+            ConfigGenerator.GeneratedValueAndExpectation generated = ConfigGenerator.generate(TenantConfig.class);
+            boolean isOk = ConfigGenerator.isOk(generated.expectation);
+
+            try {
+                TenantConfig tenantConfig = (TenantConfig) generated.value;
+                Multitenancy.addNewOrUpdateAppOrTenant(process.getProcess(), new TenantIdentifier(null, null, null), tenantConfig);
+                MultitenancyHelper.getInstance(process.getProcess()).loadConfig();
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/multitenant/RandomConfigTest.java:84]:
```suggestion
```

**Review by rishabhpoddar — APPROVED** [review]

**Pull request merged**